src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureDialog.kt

```diff
@@ -0,0 +1,342 @@
+import org.rust.lang.core.psi.RsFunction
+import org.rust.lang.core.psi.RsPsiFactory
+import org.rust.lang.core.psi.RsTypeReferenceCodeFragment
+import org.rust.lang.core.psi.ext.RsElement
+import org.rust.lang.core.psi.ext.returnType
+import org.rust.lang.core.types.ty.Ty
+import org.rust.lang.core.types.ty.TyUnit
+import org.rust.lang.core.types.type
+import org.rust.openapiext.document
+import java.awt.BorderLayout
+import java.awt.GridBagConstraints
+import java.awt.GridBagLayout
+import java.awt.Insets
+import javax.swing.JComponent
+import javax.swing.JLabel
+import javax.swing.JPanel
+
+typealias ChangeFunctionSignatureMock = (config: RsFunctionSignatureConfig) -> Unit
+
+private var MOCK: ChangeFunctionSignatureMock? = null
+
+fun showChangeFunctionSignatureDialog(
+    project: Project,
+    config: RsFunctionSignatureConfig
+) {
+    if (isUnitTestMode) {
+        val mock = MOCK ?: error("You should set mock UI via `withMockChangeFunctionSignature`")
+        mock(config)
+        RsChangeSignatureProcessor(project, config.createChangeInfo()).run()
+
+    } else {
+        ChangeSignatureDialog(project, SignatureDescriptor(config)).show()
+    }
+}
+
+@TestOnly
+fun withMockChangeFunctionSignature(mock: ChangeFunctionSignatureMock, action: () -> Unit) {
+    MOCK = mock
+    try {
+        action()
+    } finally {
+        MOCK = null
+    }
+}
+
+class SignatureParameter(val factory: RsPsiFactory, val parameter: Parameter) : ParameterInfo {
+    override fun getName(): String = parameter.patText
+    override fun getOldIndex(): Int = parameter.index
+    override fun getDefaultValue(): String? = null
+    override fun setName(name: String?) {
+        if (name != null) {
+            val pat = factory.tryCreatePat(name) ?: return
+            parameter.pat = pat
+        }
+    }
+
+    override fun getTypeText(): String = parameter.type.render(
+        includeLifetimeArguments = true,
+        includeTypeArguments = true
+    )
+
+    override fun isUseAnySingleVariable(): Boolean = false
+    override fun setUseAnySingleVariable(b: Boolean) {}
+}
+
+class SignatureDescriptor(val config: RsFunctionSignatureConfig) : MethodDescriptor<SignatureParameter, String> {
+    val function: RsFunction = config.function
+
+    override fun getName(): String = config.name
+
+    override fun getParameters(): MutableList<SignatureParameter> {
+        val factory = RsPsiFactory(config.function.project)
+
+        return config.parameters.map {
+            SignatureParameter(factory, it)
+        }.toMutableList()
+    }
+
+    override fun getParametersCount(): Int = config.parameters.size // TODO: self
+    override fun getMethod(): PsiElement = config.function
+
+    override fun getVisibility(): String = ""
+
+    /**
+     * This needs to be false, because the default dialog only offers combo boxes for visibility, but we need
+     * arbitrary strings.
+     */
+    override fun canChangeVisibility(): Boolean = false
```

Comment:
The problem is that the visibility of the function that is being refactored might not be one of these four values. Therefore, if I started to change the signature of e.g. `pub(in super::foo) fn foo()`, I could not build the dialog in such a way that it would represent the current state of the function.

In other words, if I used `Change signature`, then e.g. changed the name and executed the refactoring, it would change the visibility. Which is what I wanted to avoid (even though I agree that a combobox would be much nicer.

Reply:
Maybe it is possible to create combobox with those variants plus current visibility of function (if it not covered by variants)?

Reply:
That would be a possible solution, yes. Is there something like a combobox combined with a text field available in the platform/plugin code already? That would be even better.

Reply:
Maybe [`com.intellij.openapi.ui.ComboBox`](https://jetbrains.github.io/ui/controls/combo_box/) with `comboBox.setEditable(true)`

Reply:
Sigh, I swear that creating this small custom UI takes more time than half of the whole refactoring :D

Reply:
I used the editable combobox. Do you know if there is a way to embed a Rust code fragment into it, to get visibility auto-completion?

I can embed it into `EditorTextField`, because it has a notion of the IntelliJ `Document`, but the editor of the combobox is a Swing component that of course doesn't know anything about IntelliJ Documents.

Reply:
Oh, I don't know how to do this. I think it is ok to not have autocompletion there, because complex restricted visibility (not `pub(super)` or `pub(crate)`) are extremely rare